### ISSUE TITLE
SIGINT or Ctrl+c have different outcomes when stopping server being run via puma and rails s

### ISSUE THREAD

**Issue by shayonj** [puma/puma#1362]
### Steps to reproduce

1) Clone a simple/barebones rails app: https://github.com/shayonj/test-repro-app

2) `rails s`

3) <kbd>ctrl</kbd> + <kbd>c</kbd>

### Expected behavior

```
rails s
=> Booting Puma
=> Rails 5.1.2 application starting in development on http://localhost:3000
=> Run `rails server -h` for more startup options
Puma starting in single mode...
* Version 3.9.1 (ruby 2.3.1-p112), codename: Private Caller
* Min threads: 5, max threads: 5
* Environment: development
* Listening on tcp://0.0.0.0:3000
Use Ctrl-C to stop
^C- Gracefully stopping, waiting for requests to finish
=== puma shutdown: 2017-07-10 20:45:22 -0700 ===
- Goodbye!
```
Notice, its gracefully stopping and logs shutdown timestamp.

### Actual behavior

```
rails s
=> Booting Puma
=> Rails 5.1.2 application starting in development on http://localhost:3000
=> Run `rails server -h` for more startup options
Puma starting in single mode...
* Version 3.9.1 (ruby 2.3.1-p112), codename: Private Caller
* Min threads: 5, max threads: 5
* Environment: development
* Listening on tcp://0.0.0.0:3000
Use Ctrl-C to stop
^CExiting
```

Perhaps, no graceful stop/exit?

### System configuration

**Ruby version**: `ruby 2.3.1p112 (2016-04-26 revision 54768) [x86_64-darwin16]`
**Rails version**: `Rails 5.1.2`

### Further info
The graceful stop/exit, works fine from the cli version. For example:

1) Clone a simple/barebones rails app: https://github.com/shayonj/test-repro-app

2) `puma`

3) <kbd>ctrl</kbd> + <kbd>c</kbd>

 ```
 puma
Puma starting in single mode...
* Version 3.9.1 (ruby 2.3.1-p112), codename: Private Caller
* Min threads: 5, max threads: 5
* Environment: development
* Listening on tcp://0.0.0.0:3000
Use Ctrl-C to stop
^C- Gracefully stopping, waiting for requests to finish
=== puma shutdown: 2017-07-10 20:47:45 -0700 ===
- Goodbye!
```

----

### Proposal

This is something, I came across when working on https://github.com/puma/puma/issues/1183. After digging a little deeper, it appears [`Rails::Server` inherits from `Rack::Server`](https://github.com/rails/rails/blob/master/railties/lib/rails/commands/server/server_command.rb#L44), which traps `SIGINT` signal and `exit`s when no [`shutdown` function is found](https://github.com/rack/rack/blob/5c36acac4d298dcf04c58a10014c849a478da53f/lib/rack/server.rb#L289).

So, to better handle `SIGINT` / <kbd>ctrl</kbd> + <kbd>c</kbd>, does introducing a `.shutdown` function to [`Rack::Handler::Puma`](https://github.com/puma/puma/blob/master/lib/rack/handler/puma.rb#L5), which would restore this signal and then kill the process sound like a better workaround?

Ex:

```ruby
      def self.shutdown
        Signal.trap("SIGINT", "DEFAULT")
        Process.kill("INT", Process.pid)
      end
```

I can see an argument being made that this implementation is very specific to support Rack's requirement of a server instance having a `shutdown` function. But, given this may live in `Rack::Handler::Puma`, it still makes sense to me. Please, let me know if I am missing something :).

**Comment by shayonj** [comment]
Even the proposed solution above could work, proposed trapping of `SIGINT` as part of setting up signals in Puma's `Launcher` class at: https://github.com/puma/puma/pull/1377

**Comment by nateberkopec** [comment]
Closed by #1377
